### PR TITLE
1.31 Support

### PR DESCRIPTION
**Pull request by roydejong**
This PR adds support for Beat Saber 1.31:
- adds `DisableSsl` patch to bypass SSL / encryption / certificate checks on Ignorance/ENet connections
- minor code change for 1.31 compatibility (`environmentInfos`)
- support `use_ssl`/`useSsl` in multiplayer status data, add `MpStatusRepository` to allow BeatTogether to access this
- bumps MpCore version to 1.5

### CODE REVIEW

File: MultiplayerCore/Patchers/NetworkConfigPatcher.cs [hunk 1]
```diff
@@ -149,10 +156,17 @@ private void GetForceGameLift(ref bool __result)
...
+        }
+        
+        [AffinityPrefix]
+        [AffinityPatch(typeof(IgnoranceClient), "Start")]
+        private void PrefixIgnoranceClientStart(IgnoranceClient __instance)
+        {
+            if (!DisableSsl)
+                return;
+            
+            __instance.UseSsl = false;
```
Comment (rcelyte) [inline MultiplayerCore/Patchers/NetworkConfigPatcher.cs:168]: This completely cuts out the DTLS portion of the packets, right? I can't connect to either Official or BeatUpServer with `UseSsl = false`.

Comment (roydejong) [reply MultiplayerCore/Patchers/NetworkConfigPatcher.cs:168]: Yes, correct. When `DisableSsl` is set, packets are sent and received by the game without any encryption (straight ENet without DTLS).

The clear benefit for us is that a normal unmodified version of Ignorance/ENet on the server side "just works" in `DisableSsl` mode.

There's a related PR https://github.com/BeatTogether/BeatTogether/pull/44 for the BeatTogether plugin that allows each server config to provide this setting, and I'll follow up on the server browser side of things as well.

Comment (rcelyte) [reply MultiplayerCore/Patchers/NetworkConfigPatcher.cs:168]: Wouldn't it be better to make the option part of the status API so servers can self-report their preferred mode (and change it at will without requiring everyone to fix their configs)?

Comment (rcelyte) [reply MultiplayerCore/Patchers/NetworkConfigPatcher.cs:168]: Update: BeatUpServer now supports both encrypted and unencrypted connections by probing the type from the first packet it sees, so this isn't as much of an issue anymore (although I still personally don't like the idea of sending MultiplayerChat's voice packets unmasked over the wire).

Comment (roydejong) [reply MultiplayerCore/Patchers/NetworkConfigPatcher.cs:168]: Thanks. I agree with your feedback and I'm definitely not here to argue against or discourage encryption. I'm preparing some updates now.

**Comment by roydejong** [comment]
I've made some changes to this and the BeatTogether client PR to address feedback re: encryption and settings.

**Regarding the SSL setting:**
MultiplayerCore / BeatTogether:
- You can now pass a `use_ssl` bool in multiplayer status data (default: false). MultiplayerCore stores this but cannot enforce it, that is currently the responsibility of mods that select/override servers like BeatTogether.
- MultiplayerCore now has a `MpStatusRepository` utility that allows other mods like BeatTogether to practically access this information and receive update events.
- BeatTogether will now automatically update its configuration based on the multiplayer status data. If your master server reports `"use_ssl":true` then BeatTogether will remember and apply that setting.

ServerBrowser, unrelated to these changes:
- Will always detect & report encryption mode per lobby, and apply it when connecting.

**Comment by roydejong** [comment]
> I take issue with having defaults incompatible with Official servers. Protocol extensions like this should build on the source seamlessly, not requiring hard-coded edge cases like `if(isOfficial)` where things defer.

I don't disagree from a purist perspective. But pragmatically speaking, this default makes the most sense to me for third-party servers as I expect they will generally not do DTLS.

Extended server status data is not used for official servers at all, because all mods do indeed have special handling for official. I think "if official, disable modded extensions" type logic will always exist and is not unreasonable.

Flipping the default would have no effect except that it might be "more correct" and would require explicit opt-out from modded servers.

**Comment by rcelyte** [comment]
Modded servers already *have* an explicit opt-out. The only status responses not containing a `use_ssl` field would be servers that either expect vanilla behaviour or never supported ENet in the first place (i.e. outdated instances).

**Comment by aaronjamt** [comment]
I've further updated this PR to 1.34, when will this be merged? I don't want to submit my own PR and take @roydejong's credit for their code until this PR has been merged.

**Comment by michael-r-elp** [comment]
@roydejong would it be fine if we merge this into our dev branch instead of main, then we can open a new PR for 1.34.0

**Comment by roydejong** [comment]
Sounds good, merging into dev

**Pull request merged**